src/operators/products.c

```diff
@@ -0,0 +1,221 @@
+      //    for all vertex j in g1
+      for (j = 0; j < vcountg1; ++j) {
+         VECTOR(edges)[edge_index++] = j * vcountg2 + from; // ((j, from))
+         VECTOR(edges)[edge_index++] = j * vcountg2 + to;   // ((j, to))
+      }
+   }
+
+   IGRAPH_CHECK(igraph_create(res, &edges, vcount, directed));
+   igraph_vector_int_destroy(&edges);
+   IGRAPH_FINALLY_CLEAN(1);
+
+   return IGRAPH_SUCCESS;
+}
+
+static igraph_error_t tensor_product(igraph_t *res,
+   const igraph_t *g1, const igraph_t *g2) {
+      
+   igraph_bool_t directed1 = igraph_is_directed(g1);
+   igraph_bool_t directed2 = igraph_is_directed(g2);
+
+   if (directed1 != directed2) {
+      IGRAPH_ERROR("Tensor product between a directed and an undirected graph is invalid. "
+         "Please ensure that both graphs have the same directionality.", IGRAPH_EINVAL);
+   }
+
+   igraph_bool_t directed = directed1 && directed2;
+
+   igraph_integer_t vcountg1 = igraph_vcount(g1);
+   igraph_integer_t vcountg2 = igraph_vcount(g2);
+   igraph_integer_t ecountg1 = igraph_ecount(g1);
+   igraph_integer_t ecountg2 = igraph_ecount(g2);
+   igraph_integer_t vcount;
+   igraph_integer_t ecount;
+
+   IGRAPH_SAFE_MULT(vcountg1, vcountg2, &vcount);
+   igraph_vector_int_t edges;
+
+   // new edge count = 2*e1*e2 if undirected else e1*e2
+   IGRAPH_SAFE_MULT(ecountg1, ecountg2, &ecount);
+   if (!directed) { // directed tensor product has only e1*e2 edges, see below
+      IGRAPH_SAFE_MULT(ecount, 2, &ecount);
+   }
+   IGRAPH_SAFE_MULT(ecount, 2, &ecount);
+   IGRAPH_VECTOR_INT_INIT_FINALLY(&edges, ecount);
+
+   igraph_integer_t from1, to1, from2, to2;
+   igraph_integer_t i, j;
+   
+   // Vertex ((i, j)) with i from g1, and j from g2
+   //   will have new vertex id: i * vcountg2 + j
+   igraph_integer_t edge_index = 0;
+
+   for (i = 0; i < ecountg1; ++i) {
+      from1 = IGRAPH_FROM(g1, i);
+      to1 = IGRAPH_TO(g1, i);
+
+      for (j = 0; j < ecountg2; ++j) {
+         from2 = IGRAPH_FROM(g2, j);
+         to2 = IGRAPH_TO(g2, j);
+
+         // create edge between ((from1, from2)) to ((to1, to2))
+         VECTOR(edges)[edge_index++] = from1 * vcountg2 + from2; // ((from1, from2))
+         VECTOR(edges)[edge_index++] = to1   * vcountg2 + to2;   // ((to1, to2))
+
+         // this cross edge is not present in directed edge
+         // as to2 is not adjacent to from2, if direction is taken in account
+         if (!directed) {
+            VECTOR(edges)[edge_index++] = from1 * vcountg2 + to2;   // ((from1, to2))
+            VECTOR(edges)[edge_index++] = to1   * vcountg2 + from2; // ((to1, from2))
+         }
+      }
+   }
+
+   IGRAPH_CHECK(igraph_create(res, &edges, vcount, directed));
+   igraph_vector_int_destroy(&edges);
+   IGRAPH_FINALLY_CLEAN(1);
+
+   return IGRAPH_SUCCESS;
+}
+
+/**
+ * \function igraph_product
+ * \brief Computes the graph product of two graphs, based on the specified product type.
+ * </para><para>
```

Comment:
Be sure to have an empty line before paragraphs so that the _code_ version is readable.

Please add `\experimental` at the beginning of the docs. By policy, all new functions start out as experimental.